File: feishu-project-workflow/api-client.js
// 飞书项目流程管理配置API客户端
const axios = require('axios');
const fs = require('fs').promises;

class FeishuProjectAPI {
    constructor(config) {
        this.baseURL = 'https://project.feishu.cn/open_api';
        this.pluginToken = config.pluginToken;
        this.userKey = config.userKey;
        this.projectKey = config.projectKey;

        // 创建axios实例
        this.client = axios.create({
            baseURL: this.baseURL,
            headers: {
                'Content-Type': 'application/json',
                'X-PLUGIN-TOKEN': this.pluginToken,
                'X-USER-KEY': this.userKey
            },
            timeout: 30000
        });

        // 请求拦截器 - 添加幂等性UUID
        this.client.interceptors.request.use((config) => {
            if (['POST', 'PUT', 'PATCH', 'DELETE'].includes(config.method.toUpperCase())) {
                config.headers['X-IDEM-UUID'] = this.generateUUID();
            }
            console.log(`[API请求] ${config.method.toUpperCase()} ${config.url}`);
            return config;
        });

        // 响应拦截器
        this.client.interceptors.response.use(
            (response) => {
                console.log(`[API响应] 成功 - ${response.status}`);
                return response.data;
            },
            (error) => {
                console.error(`[API错误] ${error.message}`);
                if (error.response) {
                    console.error(`错误详情:`, error.response.data);
                }
                throw error;
            }
        );
    }

    generateUUID() {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
            const r = Math.random() * 16 | 0;
            const v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    }

    // 获取工作项类型列表
    async getWorkItemTypes() {
        try {
            const response = await this.client.get(`/${this.projectKey}/work_item_types`);
            return response.data;
        } catch (error) {
            console.error('获取工作项类型失败:', error);
            return null;
        }
    }

    // 获取流程模板列表
    async getTemplateList(workItemTypeKey) {
        try {
            const response = await this.client.get(
                `/${this.projectKey}/template_list/${workItemTypeKey}`
            );
            return response.data;
        } catch (error) {
            console.error('获取流程模板失败:', error);
            return null;
        }
    }

    // 获取字段列表
    async getFields(workItemTypeKey) {
        try {
            const response = await this.client.get(
                `/${this.projectKey}/field/${workItemTypeKey}`
            );
            return response.data;
        } catch (error) {
            console.error('获取字段列表失败:', error);
            return null;
        }
    }

    // 创建自定义字段
    async createCustomField(workItemTypeKey, fieldConfig) {
        try {
            const response = await this.client.post(
                `/${this.projectKey}/field/${workItemTypeKey}/create`,
                fieldConfig
            );
            return response.data;
        } catch (error) {
            console.error('创建自定义字段失败:', error);
            return null;
        }
    }

    // 更新流程配置
    async updateProcessConfig(workItemTypeKey, processConfig) {
        try {
            const response = await this.client.put(
                `/${this.projectKey}/process/${workItemTypeKey}/config`,
                processConfig
            );
            return response.data;
        } catch (error) {
            console.error('更新流程配置失败:', error);
            return null;
        }
    }

    // 创建流程节点
    async createProcessNode(workItemTypeKey, nodeConfig) {
        try {
            const response = await this.client.post(
                `/${this.projectKey}/process/${workItemTypeKey}/node`,
                nodeConfig
            );
            return response.data;
        } catch (error) {
            console.error('创建流程节点失败:', error);
            return null;
        }
    }

    // 创建流程转换规则
    async createTransition(workItemTypeKey, transitionConfig) {
        try {
            const response = await this.client.post(
                `/${this.projectKey}/process/${workItemTypeKey}/transition`,
                transitionConfig
            );
            return response.data;
        } catch (error) {
            console.error('创建流程转换失败:', error);
            return null;
        }
    }

    // 配置质量指标规则
    async configureMetrics(metricsConfig) {
        try {
            const response = await this.client.post(
                `/${this.projectKey}/metrics/configure`,
                metricsConfig
            );
            return response.data;
        } catch (error) {
            console.error('配置质量指标失败:', error);
            return null;
        }
    }

    // 批量创建字段
    async createFieldsBatch(workItemTypeKey, fields) {
        const results = [];
        for (const field of fields) {
            console.log(`创建字段: ${field.name}`);
            const result = await this.createCustomField(workItemTypeKey, field);
            results.push({ field: field.name, success: !!result });
            // 避免触发限流，添加延迟
            await this.delay(100);
        }
        return results;
    }

    // 延迟函数
    delay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// 主配置函数
async function configureWorkflow() {
    console.log('===== 飞书项目流程管理配置 =====\n');

    // 读取配置文件
    const workflowConfig = JSON.parse(
        await fs.readFile('./workflow-config.json', 'utf8')
    );

    // 读取认证配置
    let authConfig;
    try {
        authConfig = JSON.parse(
            await fs.readFile('./auth-config.json', 'utf8')
        );
    } catch (error) {
        console.error('请先创建 auth-config.json 文件配置认证信息');
        return;
    }

    // 初始化API客户端
    const api = new FeishuProjectAPI(authConfig);

    console.log('1. 获取现有工作项类型...');
    const workItemTypes = await api.getWorkItemTypes();
    if (workItemTypes) {
        console.log('工作项类型:', workItemTypes);
    }

    console.log('\n2. 获取需求工作项的流程模板...');
    const templates = await api.getTemplateList('requirement');
    if (templates) {
        console.log('流程模板:', templates);
    }

    console.log('\n3. 创建流程管理字段...');
    const allFields = [];

    // 收集所有节点的字段
    for (const node of workflowConfig.processManagement.nodes) {
        for (const field of node.fields) {
            // 避免重复
            if (!allFields.find(f => f.key === field.key)) {
                allFields.push({
                    key: field.key,
                    name: field.name,
                    type: field.type,
                    required: field.required || false,
                    description: field.description || '',
                    options: field.options || null,
                    default: field.default || null
                });
            }
        }
    }

    console.log(`准备创建 ${allFields.length} 个字段`);
    const fieldResults = await api.createFieldsBatch('requirement', allFields);

    console.log('\n字段创建结果:');
    fieldResults.forEach(result => {
        console.log(`  - ${result.field}: ${result.success ? '✅ 成功' : '❌ 失败'}`);
    });

    console.log('\n4. 创建流程节点...');
    for (const node of workflowConfig.processManagement.nodes) {
        console.log(`创建节点: ${node.name}`);
        const nodeResult = await api.createProcessNode('requirement', {
            id: node.id,
            name: node.name,
            type: node.type,
            fields: node.fields.map(f => f.key)
        });

        if (nodeResult) {
            console.log(`  ✅ ${node.name} 创建成功`);
        } else {
            console.log(`  ❌ ${node.name} 创建失败`);
        }

        await api.delay(100);
    }

    console.log('\n5. 创建流程转换规则...');
    for (const transition of workflowConfig.processManagement.transitions) {
        console.log(`创建转换: ${transition.name}`);
        const transitionResult = await api.createTransition('requirement', transition);

        if (transitionResult) {
            console.log(`  ✅ ${transition.name} 创建成功`);
        } else {
            console.log(`  ❌ ${transition.name} 创建失败`);
        }

        await api.delay(100);
    }

    console.log('\n6. 配置质量指标...');
    const metricsResult = await api.configureMetrics(workflowConfig.qualityMetrics);
    if (metricsResult) {
        console.log('✅ 质量指标配置成功');
    } else {
        console.log('❌ 质量指标配置失败');
    }

    console.log('\n===== 配置完成 =====');

    // 生成配置报告
    const report = {
        timestamp: new Date().toISOString(),
        project: authConfig.projectKey,
        fields: {
            total: allFields.length,
            created: fieldResults.filter(r => r.success).length,
            failed: fieldResults.filter(r => !r.success).length
        },
        nodes: workflowConfig.processManagement.nodes.length,
        transitions: workflowConfig.processManagement.transitions.length,
        metrics: Object.keys(workflowConfig.qualityMetrics).length
    };

    await fs.writeFile(
        './configuration-report.json',
        JSON.stringify(report, null, 2)
    );

    console.log('\n配置报告已保存到 configuration-report.json');
}

// 导出模块
module.exports = {
    FeishuProjectAPI,
    configureWorkflow
};

// 如果直接运行脚本
if (require.main === module) {
    configureWorkflow().catch(console.error);
}

File: feishu-project-workflow/api_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
飞书项目流程管理配置API客户端 (Python版本)
"""

import json
import time
import uuid
from typing import Dict, List, Optional, Any
import requests
from datetime import datetime


class FeishuProjectAPI:
    """飞书项目API客户端"""

    def __init__(self, config: Dict[str, str]):
        """初始化API客户端

        Args:
            config: 包含 pluginToken, userKey, projectKey 的配置字典
        """
        self.base_url = 'https://project.feishu.cn/open_api'
        self.plugin_token = config['pluginToken']
        self.user_key = config['userKey']
        self.project_key = config['projectKey']

        # 创建会话
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'X-PLUGIN-TOKEN': self.plugin_token,
            'X-USER-KEY': self.user_key
        })

    def _generate_uuid(self) -> str:
        """生成幂等性UUID"""
        return str(uuid.uuid4())

    def _request(self, method: str, endpoint: str, data: Optional[Dict] = None) -> Optional[Dict]:
        """发送API请求

        Args:
            method: HTTP方法
            endpoint: API端点
            data: 请求数据

        Returns:
            响应数据或None（如果请求失败）
        """
        url = f"{self.base_url}/{self.project_key}/{endpoint}"

        # 为写操作添加幂等性UUID
        headers = {}
        if method in ['POST', 'PUT', 'PATCH', 'DELETE']:
            headers['X-IDEM-UUID'] = self._generate_uuid()

        try:
            print(f"[API请求] {method} {endpoint}")
            response = self.session.request(
                method=method,
                url=url,
                json=data,
                headers=headers,
                timeout=30
            )

            if response.status_code == 200:
                result = response.json()
                if result.get('err_code') == 0:
                    print(f"[API响应] 成功")
                    return result.get('data')
                else:
                    print(f"[API错误] {result.get('err_msg')}")
                    return None
            else:
                print(f"[HTTP错误] {response.status_code}: {response.text}")
                return None

        except requests.exceptions.RequestException as e:
            print(f"[请求异常] {str(e)}")
            return None

    def get_work_item_types(self) -> Optional[List]:
        """获取工作项类型列表"""
        return self._request('GET', 'work_item_types')

    def get_template_list(self, work_item_type_key: str) -> Optional[List]:
        """获取流程模板列表"""
        return self._request('GET', f'template_list/{work_item_type_key}')

    def get_fields(self, work_item_type_key: str) -> Optional[List]:
        """获取字段列表"""
        return self._request('GET', f'field/{work_item_type_key}')

    def create_custom_field(self, work_item_type_key: str, field_config: Dict) -> Optional[Dict]:
        """创建自定义字段"""
        return self._request('POST', f'field/{work_item_type_key}/create', field_config)

    def update_process_config(self, work_item_type_key: str, process_config: Dict) -> Optional[Dict]:
        """更新流程配置"""
        return self._request('PUT', f'process/{work_item_type_key}/config', process_config)

    def create_process_node(self, work_item_type_key: str, node_config: Dict) -> Optional[Dict]:
        """创建流程节点"""
        return self._request('POST', f'process/{work_item_type_key}/node', node_config)

    def create_transition(self, work_item_type_key: str, transition_config: Dict) -> Optional[Dict]:
        """创建流程转换规则"""
        return self._request('POST', f'process/{work_item_type_key}/transition', transition_config)

    def configure_metrics(self, metrics_config: Dict) -> Optional[Dict]:
        """配置质量指标"""
        return self._request('POST', 'metrics/configure', metrics_config)

    def create_fields_batch(self, work_item_type_key: str, fields: List[Dict]) -> List[Dict]:
        """批量创建字段"""
        results = []
        for field in fields:
            print(f"创建字段: {field['name']}")
            result = self.create_custom_field(work_item_type_key, field)
            results.append({
                'field': field['name'],
                'success': result is not None
            })
            # 避免触发限流
            time.sleep(0.1)
        return results


def configure_workflow():
    """主配置函数"""
    print("===== 飞书项目流程管理配置 =====\n")

    # 读取配置文件
    try:
        with open('workflow-config.json', 'r', encoding='utf-8') as f:
            workflow_config = json.load(f)
    except FileNotFoundError:
        print("错误: 找不到 workflow-config.json 文件")
        return

    # 读取认证配置
    try:
        with open('auth-config.json', 'r', encoding='utf-8') as f:
            auth_config = json.load(f)
    except FileNotFoundError:
        print("错误: 请先创建 auth-config.json 文件配置认证信息")
        print("可以复制 auth-config-template.json 并填入您的认证信息")
        return

    # 初始化API客户端
    api = FeishuProjectAPI(auth_config)

    print("1. 获取现有工作项类型...")
    work_item_types = api.get_work_item_types()
    if work_item_types:
        print(f"工作项类型: {json.dumps(work_item_types, ensure_ascii=False, indent=2)}")

    print("\n2. 获取需求工作项的流程模板...")
    templates = api.get_template_list('requirement')
    if templates:
        print(f"流程模板: {json.dumps(templates, ensure_ascii=False, indent=2)}")

    print("\n3. 创建流程管理字段...")
    all_fields = []

    # 收集所有节点的字段
    for node in workflow_config['processManagement']['nodes']:
        for field in node['fields']:
            # 避免重复
            if not any(f['key'] == field['key'] for f in all_fields):
                all_fields.append({
                    'key': field['key'],
                    'name': field['name'],
                    'type': field['type'],
                    'required': field.get('required', False),
                    'description': field.get('description', ''),
                    'options': field.get('options'),
                    'default': field.get('default')
                })

    print(f"准备创建 {len(all_fields)} 个字段")
    field_results = api.create_fields_batch('requirement', all_fields)

    print("\n字段创建结果:")
    for result in field_results:
        status = "✅ 成功" if result['success'] else "❌ 失败"
        print(f"  - {result['field']}: {status}")

    print("\n4. 创建流程节点...")
    for node in workflow_config['processManagement']['nodes']:
        print(f"创建节点: {node['name']}")
        node_result = api.create_process_node('requirement', {
            'id': node['id'],
            'name': node['name'],
            'type': node['type'],
            'fields': [f['key'] for f in node['fields']]
        })

        if node_result:
            print(f"  ✅ {node['name']} 创建成功")
        else:
            print(f"  ❌ {node['name']} 创建失败")

        time.sleep(0.1)

    print("\n5. 创建流程转换规则...")
    for transition in workflow_config['processManagement']['transitions']:
        print(f"创建转换: {transition['name']}")
        transition_result = api.create_transition('requirement', transition)

        if transition_result:
            print(f"  ✅ {transition['name']} 创建成功")
        else:
            print(f"  ❌ {transition['name']} 创建失败")

        time.sleep(0.1)

    print("\n6. 配置质量指标...")
    metrics_result = api.configure_metrics(workflow_config['qualityMetrics'])
    if metrics_result:
        print("✅ 质量指标配置成功")
    else:
        print("❌ 质量指标配置失败")

    print("\n===== 配置完成 =====")

    # 生成配置报告
    report = {
        'timestamp': datetime.now().isoformat(),
        'project': auth_config['projectKey'],
        'fields': {
            'total': len(all_fields),
            'created': sum(1 for r in field_results if r['success']),
            'failed': sum(1 for r in field_results if not r['success'])
        },
        'nodes': len(workflow_config['processManagement']['nodes']),
        'transitions': len(workflow_config['processManagement']['transitions']),
        'metrics': len(workflow_config['qualityMetrics'])
    }

    with open('configuration-report.json', 'w', encoding='utf-8') as f:
        json.dump(report, f, ensure_ascii=False, indent=2)

    print("\n配置报告已保存到 configuration-report.json")


if __name__ == '__main__':
    configure_workflow()